File: bdc.js

const { Client } = require('node-rfc');

// SAP 연결 설정
const abapSystem = {
    user: 'SAP_USER',
    passwd: 'SAP_PASSWORD',
    ashost: 'SAP_SERVER',
    sysnr: 'SYSNR',
    client: 'CLIENT',
    lang: 'EN'
};

const client = new Client(abapSystem);

async function callTransactionUsing() {
    try {
        // SAP 시스템에 연결
        await client.open();

        // BDC 데이터 구성
        const lt_bdcdata = [
            { PROGRAM: 'SAPL...', DYNPRO: '0100', DYNBEGIN: 'X' },
            { FNAM: 'BDC_OKCODE', FVAL: '/00' },
            { FNAM: 'RFBELN', FVAL: '123456' }
        ];

        // RFC 호출
        const result = await client.call('RFC_CALL_TRANSACTION_USING', {
            TCODE: 'VA01',
            MODE: 'E',
            UPDATE: 'S',
            BDCTAB: lt_bdcdata
        });

        if (result.MESSAGES) {
            const lt_messages = result.MESSAGES;
            lt_messages.forEach(message => {
                console.log(message);
            });
        } else {
            console.log('Transaction successful');
        }
    } catch (err) {
        console.error(`Error: ${err.message}`);
    } finally {
        // SAP 연결 종료
        await client.close();
    }
}

callTransactionUsing();

File: Test/bdc.py
import pyrfc
from pyrfc import Connection

# SAP 연결 설정
conn = Connection(
    user='SAP_USER',
    passwd='SAP_PASSWORD',
    ashost='SAP_SERVER',
    sysnr='SYSNR',
    client='CLIENT',
    lang='EN'
)

# BDC 데이터 구성
lt_bdcdata = [
    {'PROGRAM': 'SAPL...', 'DYNPRO': '0100', 'DYNBEGIN': 'X'},
    {'FNAM': 'BDC_OKCODE', 'FVAL': '/00'},
    {'FNAM': 'RFBELN', 'FVAL': '123456'}
]

# RFC 호출
try:
    result = conn.call('RFC_CALL_TRANSACTION_USING', 
                       TCODE='VA01', 
                       MODE='E', 
                       UPDATE='S', 
                       BDCTAB=lt_bdcdata)
    
    if 'MESSAGES' in result:
        lt_messages = result['MESSAGES']
        for message in lt_messages:
            print(message)
    else:
        print("Transaction successful")

except pyrfc.CommunicationError as e:
    print(f"Communication error: {e}")
except pyrfc.LogonError as e:
    print(f"Logon error: {e}")
except Exception as e:
    print(f"An error occurred: {e}")

